client/client-rulesengine/src/main/java/software/amazon/smithy/java/client/rulesengine/RulesProgram.java

```diff
@@ -0,0 +1,408 @@
+     * byte that defines the number of elements in the list.
+     */
+    static final byte CREATE_LIST = 10;
+
+    /**
+     * Pops N*2 values off the stack (key then value), creates a map of those values, and pushes the map onto the
+     * stack. Each popped key must be a string. Must be followed by an unsigned byte that defines the
+     * number of entries in the map.
+     */
+    static final byte CREATE_MAP = 11;
+
+    /**
+     * Resolves a template string. Must be followed by two bytes, a short, that represents the constant pool index
+     * that stores the StringTemplate.
+     *
+     * <p>The corresponding instruction has a StringTemplate that tells the VM how many values to pop off the stack.
+     * The popped values fill in values into the template. The resolved template value as a string is then pushed onto
+     * the stack.
+     */
+    static final byte RESOLVE_TEMPLATE = 12;
+
+    /**
+     * Calls a function. Must be followed by a byte to provide the function index to call.
+     *
+     * <p>The function pops zero or more values off the stack based on the RulesFunction registered for the index,
+     * and then pushes the Object result onto the stack.
+     */
+    static final byte FN = 13;
+
+    /**
+     * Pops the top level value and applies a getAttr expression on it, pushing the result onto the stack.
+     *
+     * <p>Must be followed by two bytes, a short, that represents the constant pool index that stores the
+     * AttrExpression.
+     */
+    static final byte GET_ATTR = 14;
+
+    /**
+     * Pops a value and pushes true if the value is boolean true, false if not.
+     */
+    static final byte IS_TRUE = 15;
+
+    /**
+     * Checks if a register is boolean true and pushes the result onto the stack.
+     *
+     * <p>Must be followed by a byte that represents the register to check.
+     */
+    static final byte TEST_REGISTER_IS_TRUE = 16;
+
+    /**
+     * Pops the value at the top of the stack and returns it from the VM. This can be used for testing purposes or
+     * for returning things other than endpoint values.
+     */
+    static final byte RETURN_VALUE = 17;
+
+    final List<RulesExtension> extensions;
+    final Object[] constantPool;
+    final byte[] instructions;
+    final int instructionOffset;
+    final int instructionSize;
+    final ParamDefinition[] registerDefinitions;
+    final RulesFunction[] functions;
+    private final BiFunction<String, Context, Object> builtinProvider;
+    private final int paramCount; // number of provided params.
+
+    RulesProgram(
+            List<RulesExtension> extensions,
+            byte[] instructions,
+            int instructionOffset,
+            int instructionSize,
+            List<ParamDefinition> params,
+            RulesFunction[] functions,
+            BiFunction<String, Context, Object> builtinProvider,
+            final Object[] constantPool
```

Comment:
Any reason for this one in particular to be `final`?

Reply:
nope. fixed